File: src/routes/assessment-section/assessment-section.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
} from '@nestjs/common'
import { AssessmentSection } from '@prisma/client'
import { AuthType } from '../../shared/constants/auth.constant'
import { RoleName } from '../../shared/constants/role.constant'
import { Auth } from '../../shared/decorators/auth.decorator'
import { Roles } from '../../shared/decorators/roles.decorator'
import {
  CreateAssessmentSectionDto,
  UpdateAssessmentSectionDto,
  AssessmentSectionQueryDto,
} from './assessment-section.dto'
import type {
  AssessmentSectionBasic,
  AssessmentSectionQuery,
  AssessmentSectionWithItems,
} from './assessment-section.model'
import { AssessmentSectionService } from './assessment-section.service'

@Controller('assessment-sections')
export class AssessmentSectionController {
  constructor(private readonly assessmentSectionService: AssessmentSectionService) {}

  @Post()
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Admin)
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() createDto: CreateAssessmentSectionDto): Promise<AssessmentSection> {
    return this.assessmentSectionService.createAssessmentSection({
      assessmentId: createDto.assessmentId,
      title: createDto.title,
      type: createDto.type,
      timeLimitSec: createDto.timeLimitSec,
    })
  }

  @Get()
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Customer, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async findAll(@Query() queryDto: AssessmentSectionQueryDto): Promise<{
    data: AssessmentSectionBasic[]
    pagination: {
      total: number
      page: number
      limit: number
      totalPages: number
      hasNext: boolean
      hasPrev: boolean
    }
  }> {
    return this.assessmentSectionService.getAssessmentSections(queryDto as AssessmentSectionQuery)
  }

  @Get(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Customer, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async findOne(@Param('id', ParseIntPipe) id: number): Promise<AssessmentSection> {
    return this.assessmentSectionService.getAssessmentSection(id)
  }

  @Get(':id/items')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Customer, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async getItems(@Param('id', ParseIntPipe) id: number): Promise<AssessmentSectionWithItems> {
    return this.assessmentSectionService.getAssessmentSectionWithItems(id)
  }

  @Put(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async update(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateDto: UpdateAssessmentSectionDto,
  ): Promise<AssessmentSection> {
    return this.assessmentSectionService.updateAssessmentSection(id, updateDto)
  }

  @Delete(':id')
  @Auth([AuthType.Bearer])
  @Roles(RoleName.Staff, RoleName.Lecturer, RoleName.Admin)
  @HttpCode(HttpStatus.OK)
  async remove(@Param('id', ParseIntPipe) id: number) {
    return this.assessmentSectionService.deleteAssessmentSection(id)
  }
}

File: assessment-service/src/main/java/com/torii/assessment/controller/AssessmentSectionController.java
package com.torii.assessment.controller;

import com.torii.assessment.dto.section.*;
import com.torii.assessment.service.AssessmentSectionService;
import com.torii.assessment.util.RequestAuthUtil;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/assessment-sections")
@RequiredArgsConstructor
@Tag(name = "Assessment Sections", description = "Assessment section management APIs")
public class AssessmentSectionController {

    private final AssessmentSectionService assessmentSectionService;

    @PostMapping
    @Operation(summary = "Create a new assessment section")
    public ResponseEntity<AssessmentSectionDTO> create(@Valid @RequestBody CreateAssessmentSectionDTO dto,
                                                       HttpServletRequest request) {
        Integer userId = RequestAuthUtil.requireUser(request).userId();
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(assessmentSectionService.createAssessmentSection(dto, userId));
    }

    @GetMapping
    @Operation(summary = "Get all assessment sections with pagination and filters")
    public ResponseEntity<AssessmentSectionListResponseDTO> findAll(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer limit,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String type,
            @RequestParam(required = false) Long assessmentId,
            @RequestParam(required = false, defaultValue = "id") String sortBy,
            @RequestParam(required = false, defaultValue = "asc") String sortOrder) {

        QueryAssessmentSectionDTO queryDto = QueryAssessmentSectionDTO.builder()
                .page(page != null ? page : 1)
                .limit(limit != null ? limit : 20)
                .search(search)
                .type(type)
                .assessmentId(assessmentId)
                .sortBy(sortBy)
                .sortOrder(sortOrder)
                .build();

        return ResponseEntity.ok(assessmentSectionService.getAssessmentSections(queryDto));
    }

    @GetMapping("/{id}")
    @Operation(summary = "Get assessment section by ID")
    public ResponseEntity<AssessmentSectionDTO> findOne(@PathVariable Long id) {
        return ResponseEntity.ok(assessmentSectionService.getAssessmentSection(id));
    }

    @GetMapping("/{id}/items")
    @Operation(summary = "Get assessment section with its items")
    public ResponseEntity<AssessmentSectionWithItemsDTO> getItems(@PathVariable Long id) {
        return ResponseEntity.ok(assessmentSectionService.getAssessmentSectionWithItems(id));
    }

    @PutMapping("/{id}")
    @Operation(summary = "Update an assessment section")
    public ResponseEntity<AssessmentSectionDTO> update(
            @PathVariable Long id,
            @Valid @RequestBody UpdateAssessmentSectionDTO dto,
            HttpServletRequest request) {
        Integer userId = RequestAuthUtil.requireUser(request).userId();
        return ResponseEntity.ok(assessmentSectionService.updateAssessmentSection(id, dto, userId));
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "Delete an assessment section")
    public ResponseEntity<Map<String, String>> remove(@PathVariable Long id, HttpServletRequest request) {
        Integer userId = RequestAuthUtil.requireUser(request).userId();
        assessmentSectionService.deleteAssessmentSection(id, userId);
        return ResponseEntity.ok(Map.of("message", "Assessment section deleted successfully"));
    }
}
